[Li Bo Feng] iP

bofeng1999 opened the pull request:
## Duke 

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)



If you Java programmer, you can use it to practice Java too. Here's the ```main``` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

bofeng1999 committed:
Implement level 1

bofeng1999 committed:
Implement level 2

bofeng1999 committed:
Implement Level 3

bofeng1999 committed:
Implement Level 4

bofeng1999 committed:
Implement A-TextUiTesting

bofeng1999 committed:
Implement Handle Errors

bofeng1999 committed:
Implement Delete

bofeng1999 committed:
Fix A-TextUiTesting

bofeng1999 committed:
Implement saving and loading data

bofeng1999 committed:
Save done/undone

bofeng1999 committed:
Implement date time

bofeng1999 committed:
Merge branch 'branch-Level-8'

bofeng1999 committed:
Save date time properly

bofeng1999 committed:
Extract out Storage class

bofeng1999 committed:
Extract out TaskList

bofeng1999 committed:
Extract UI class

bofeng1999 committed:
Extract Parser class

bofeng1999 committed:
Merge branch 'add-gradle-support' of github.com:bofeng1999/ip

bofeng1999 committed:
Add gradle and organise into packages

bofeng1999 committed:
Add JUnit test for Parser

bofeng1999 committed:
Fix file path

bofeng1999 committed:
Implement find

bofeng1999 committed:
Fix coding standard

bofeng1999 committed:
Implement Java Doc

bofeng1999 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Duke.java

bofeng1999 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/ui/UI.java

bofeng1999 committed:
Fix checkstyle

TabrizPlv commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,96 @@
+        }
+
+        while (true) {
+            try {
+                String[] command = sc.nextLine().split(" ", 2);
+                if (command[0].equals("bye")) {
+                    storage.saveTasks(taskList);
+                    ui.printMessage("Bye. Hope to see you again soon!");
+                    break;
+                } else if (command[0].equals("list")) {
+                    ui.printMessage(taskList.listTasks());
+                } else if (command[0].equals("mark")) {
+                    int taskNum = parser.getTaskNum(command);
+                    taskList.markTask(parser.getTaskNum(command));
+                    ui.printSuccessMessage("Nice! I've marked this task as done:",
+                            taskList.getTask(taskNum));
+                } else if (command[0].equals("unmark")) {
+                    int taskNum = parser.getTaskNum(command);
+                    taskList.unmarkTask(taskNum);
+                    ui.printSuccessMessage("OK, I've marked this task as not done yet:",
+                            taskList.getTask(taskNum));
+                } else if (command[0].equals("delete")) {
+                    int taskNum = parser.getTaskNum(command);
+                    Task removedTask = taskList.deleteTask(taskNum);
+                    ui.printTaskMessage("Noted. I've removed this task:",
+                            removedTask, taskList.getSize());
+                } else if (command[0].equals("find")) {
+                    String keyword = parser.getKeyword(command);
+                    ui.printFindResult(taskList.getTasksWithKeyword(keyword));
+                } else {
+                    taskList.addTask(parser.getTaskToAdd(command));
+                    ui.printTaskMessage("Got it. I've added this task:",
+                            taskList.getLatestTask(), taskList.getSize());
+                }
+            } catch (DukeException e) {
+                ui.printMessage(e.getMessage());
+            } catch (NumberFormatException e) {
+                ui.printMessage("Valid task number required");
+            } catch (IOException e) {
+                ui.printMessage(e.getMessage());
```
Very sound flow of logic! But i feel that this logic can be abstracted into another class method! 😃

TabrizPlv commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,10 @@
+package duke.exception;
+
+/**
+ * Class to define exceptions specific to Duke.
```
Nice concise description! Very easy to understand

TabrizPlv commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    /**
+     * Generate the name of the task.
```
Header comments for getters and setters may not be necessary!

TabrizPlv reviewed (commented):
LGTM. Just minor improvements left.

bofeng1999 committed:
Implement GUI

bofeng1999 committed:
Update README.md

bofeng1999 committed:
Merge branch 'master' of github.com:bofeng1999/ip

bofeng1999 committed:
Fix checkstyle

axmszr commented on src/main/java/duke/ui/MainWindow.java:
```diff
@@ -0,0 +1,65 @@
+
+    private Duke duke;
+
+    private Image userImage = new Image(this.getClass().getResourceAsStream("/images/Golden.jpeg"));
+    private Image dukeImage = new Image(this.getClass().getResourceAsStream("/images/Pug.png"));
```
I love them.

axmszr commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,101 @@
+    public String getResponse(String input) {
+        String[] command = input.split(" ", 2);
+        try {
+            if (command[0].equals("bye")) {
```
Would this `if`-`else` block be more readable as a `switch`-`case` block?

axmszr commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,74 @@
+     * @return Task object specified by the user command.
+     * @throws DukeException User provide invalid command.
+     */
+
```
Should there be a blank line here?

axmszr commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,74 @@
+        switch (command[0]) {
+        case "todo":
+            return new Todo(command[1]);
+        case "deadline": {
```
I think the Java standard we follow doesn't use Egyptian {} braces for cases! Just the colon and indentation.

axmszr commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,74 @@
+        case "todo":
+            return new Todo(command[1]);
+        case "deadline": {
+            String[] arguments = command[1].split(" /by ");
```
The whitespace around the `"/by"` is a good idea!

axmszr commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,135 @@
+        }
+        Task removedTask = tasks.get(taskNum);
+        tasks.remove(taskNum);
+        return removedTask;
```
Oh! `remove` for an `ArrayList` acts like `pop`. You don't need to store the Task first!
```suggestion
        return tasks.get(taskNum);
```

axmszr reviewed (commented):
Looks great! Pretty neat and understandable code throughout, nice JavaDocs and good names.
Jiayous for iP! :D

bofeng1999 committed:
FIx JavaDoc

bofeng1999 committed:
Add Java assertions to code

There are various sections in the code base where assertions can be used
to ensure the integrity of the program.

For example, when methods take in an object, it is good to assert that
the object is not null to prevent NullPointerException.

bofeng1999 committed:
Merge pull request #1 from bofeng1999/branch-A-Assertions

Add Java assertions to code

bofeng1999 committed:
Improve code quality

There are various parts in the code where the quality is improved for
better readability. For example, instead of having a long if statement
with many else-ifs, a switch statement is used instead.

bofeng1999 committed:
Merge pull request #2 from bofeng1999/branch-A-CodeQuality

Improve code quality

bofeng1999 committed:
Implement detection of duplicated tasks

bofeng1999 committed:
Merge pull request #3 from bofeng1999/branch-C-DetectDuplicates

Implement detection of duplicated tasks

bofeng1999 committed:
Fix JavaDoc

bofeng1999 committed:
Update README.md

bofeng1999 committed:
Upload Ui.png

bofeng1999 committed:
Add User Guide

bofeng1999 committed:
Fix Javadoc

bofeng1999 committed:
Improve User Guide

bofeng1999 committed:
Update README.md

bofeng1999 committed:
Update README.md

bofeng1999 committed:
Fix Readme
